### PR TITLE
feat(tests): Add snapshot testing

### PR DESCRIPTION
**Pull request by last-partizan**
Hello, i would like to add snapshot testing for this app.

This should make easier to write and update functional tests from tests/cli.

I'm preparing rewrite of pocount, and current cli tests are incomplete. This ensures output will not change for all option combinations after refactoring.

Snapshots can be generated/updated with `pytest --snapshot-update`.

### CODE REVIEW

File: translate/tools/tests.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import pytest
+
+from syrupy.assertion import SnapshotAssertion
+
+
+po_files = """
+tests/cli/data/test_pocount_po_csv/one.po
+tests/cli/data/test_pocount_po_file/one.po
+tests/cli/data/test_pocount_po_fuzzy/one.po
+""".split()
```
Comment (nschonni) [inline translate/tools/tests.py:12]: Shouldn't this be declared as an array, instead of a set of strings that is split into an array?

**Comment by nschonni** [comment]
I'm guessing moving these tests to the existing file https://github.com/translate/translate/blob/61132d19424d97acb630b22fb52d8687c716ff35/translate/tools/test_pocount.py would get rid of some of the errors in the CI. The other error is for overriting the built-in `opts` I think

**Comment by last-partizan** [comment]
If i understand correctly, `test_pocount.po` is for unit-tests, my snapshot tests is functional tests targeted to replace `tests/cli/`.

Would be ok to put my those snapshot tests into `tests/cli/test_pocount.py`?

And probably one other `test_help.py`, to contain all `--help` checks (they're completely identical for all apps, so it we wouldn't need to repeat it for each.

**Comment by nijel** [comment]
Looks nice.

I'm uncertain if it's still reasonable to split the tests once they both use pytest. So I'd place all `pocount` tests to `translate/tools/test_pocount.py`.

There is really no separation of unit and functional tests for the rest of the code. Convertor tests (`translate/convert/test_*.py`) typically also include both unit and functional tests.

**Comment by last-partizan** [comment]
Btw, should i remove old cli tests for pocount and junitmsgfmt?

**Comment by last-partizan** [comment]
This turned out trickier then i expected.

When adding "--help" snapshots for all tools, i can see that output is different for different python versions.

Should we remove them? I think that's safe option - testing ArgumentParser should be left to it's own testsuite.

**Comment by nijel** [comment]
How does it differ between Python versions? Is optparse output different or something we can influence?

**Comment by last-partizan** [comment]
For example, python 3.7 vs 3.10 - different names for "optional arguments". Between other versions indents or text aligning may be different.

```
>       assert stdout == snapshot
E       AssertionError: assert [+ received] == [- snapshot]
E           '''
E           ...
E
E         - optional arguments:
E         + options:
E             -h, --help            show this help message and exit
E            ...
E           '''
```

**Comment by nijel** [comment]
Hmm, that really doesn't look suitable for snapshot testing.

**Comment by last-partizan** [comment]
I was thinking how it was working before, because i removed "snapshot" for old-style bash tests.

But turns out it was not checking snapshots for `--help` tests, it was only checking that it "has_stdout".

```
python $PYTHON_ARGS $(which pocount) -h
start_checks
has_stdout
end_checks
```

If you think this type of test may be useful, i can add somethink like `stdout != ""`. But i don't think it makes much  sense.

**Comment by nschonni** [comment]
Maybe https://pypi.org/project/pytest-snapshot/ would work with adding a python version to the file path `snapshot.assert_match(foo('function input'), 'help_output_37.txt')`

**Comment by last-partizan** [comment]
@nschonni it could work even with pytest-syrupy, but we still need to generate those snapshots. And i don't expect every developer would have all python versions installed.

And leaving snapshot generation to autocommit bot would defeat the purpose of having them at all (becouse the goal is to have only expected changes, and not update it automatically).

**Comment by last-partizan** [comment]
One possibility that i can imagine - is restricting those `--help` snapshots to latest version of python.

It would definitely work, but we will require for devs to use latest version of python. That's easier that all of them, but still a limiting factor.

I suggest to not use snapshots for `--help` at all. There is no way we can break argparser default functionality, so we should not test it.

We should be testing only our code - and we're already doing it.

**Comment by nijel** [comment]
Potentially, we could break optparse to not emit help, so testing that help works and contains some commands makes sense (I believe we currently have such tests). Testing full output is not really desired here as it a moving target.

**Comment by last-partizan** [comment]
Added check that stdout from `commandname --help` contains at leas command name and "--help".

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Pull request merged**